gasometer/src/lib.rs

```diff
@@ -432,26 +544,41 @@ pub fn dynamic_opcode_cost<H: Handler>(
 		Opcode::SELFBALANCE if config.has_self_balance => GasCost::Low,
 		Opcode::SELFBALANCE => GasCost::Invalid,
 
-		Opcode::EXTCODESIZE => GasCost::ExtCodeSize,
-		Opcode::BALANCE => GasCost::Balance,
+		Opcode::EXTCODESIZE => GasCost::ExtCodeSize {
+			address: stack.peek(0)?.into(),
```

Comment:
The list of `accessed_addresses` and `accessed_storages` really shouldn't belong to `gasometer`. A better design is to create a separate struct `Accessed` specifically to hold them, and add an extra field in executor metadata. This line can then be changed to a `is_cold: bool` which we continue to handle the access gas.

The EVM step cycle has two phrases, one `immutable` where we gather necessary data such as gas cost, to determine if we should continue the execute, and another `mutable` where we actually do the execution. By putting access list to gasometer this invariant is broken (which is made apparent by the below change of an immutable reference to a mutable one).

* `immutable` peek the access list, determine if the address is cold.
* `immutable` pass the coldness along with other information to gasometer and get the gas costs.
* `mutable` do the execution / update the gas metadata and access list.

The above invariant may not sound that useful if it's only about a full execution cycle, but it is extremely relevant if we want to use this crate to build a debugger or a stepper.

Reply:
I had noticed this pattern and my first attempt at this functionality went in this direction. You can see it in [this commit](https://github.com/rust-blockchain/evm/pull/40/commits/d7ea7678b288f9461949f1ac686653f649812829#diff-bf2551b9edd7333cb28398048e01898cfd8c41877603ff3e17ea444fe8cf6e51R784).

What I didn't like about this approach is the need to thread the `is_cold` functionality through the `Handler` because that was the only way for the gasomete to get access to the wider state. But that felt off to me because it is only the gasometer that cares about these access lists, so I thought it might be better to encapsulate inside the gasometer itself.

But I respect if you think maintaining this strict mutability flow is more important. Just let me know and I can change it back to my first version.

Reply:
Yeah I do think that version would be better. `Handler::is_cold` is okay -- we do that for `Handler::exists` as well (which is kind of also EIP-specific).

Reply:
Sounds good, thanks for the clarification. I'm OOO tomorrow, but I'll push a fix for this on Monday.

Reply:
I can get this is before Monday.

Reply:
Here: https://github.com/rust-blockchain/evm/pull/40/commits/c2af4516d210b7520f6cb1c4c5d2d16b6a9d6103